src/commons/utils/RequestHelper.tsx

```diff
@@ -0,0 +1,155 @@
+ *
+ * NOTE: If multiple API calls are made with an expired access token, only one refresh token flow will be made.
+ * All these API calls will then be retried thereafter.
+ */
+export const request = async (
+  path: string,
+  method: string,
+  opts: RequestOptions
+): Promise<Response | null> => {
+  const fetchOptions = generateApiCallHeadersAndFetchOptions(method, opts);
+
+  try {
+    const resp = await fetch(`${Constants.backendUrl}/v2/${path}`, fetchOptions);
+    if (resp.ok) {
+      return resp;
+    }
+
+    // Non-unauthorized errors (these don't need to trigger refresh token flow)
+    if (resp.status !== 401) {
+      showWarningMessage(
+        opts.errorMessage
+          ? opts.errorMessage
+          : `Error while communicating with backend: ${resp.status} ${resp.statusText}`
+      );
+      return null;
+    }
+
+    // Unauthorized errors (trigger refresh token flow)
+    if (!opts.refreshToken) {
+      // If no refreshToken passed in, we are in `postRefresh` itself AND the refresh token flow failed
+      // (i.e. invalid refresh token sent)
+      return null;
+    }
+
+    // Perform refresh token flow before retrying API call.
+    // If refresh token flow or retried API call 401s again, determine whether to force logout.
+    try {
+      if (!refreshingTokensPromise) {
+        refreshingTokensPromise = postRefresh(opts.refreshToken);
+      }
+
+      // Blocks other API calls that failed due to expired access token, while performing a single refresh token flow
+      const newTokens = await refreshingTokensPromise;
+      if (!newTokens) {
+        throw Error('Invalid refreshed token');
+      }
+
+      store.dispatch(actions.setTokens(newTokens));
+      fetchOptions.headers.set('Authorization', `Bearer ${newTokens.accessToken}`);
+
+      const retriedResp = await fetch(`${Constants.backendUrl}/v2/${path}`, fetchOptions);
+      if (retriedResp.ok) {
+        return retriedResp;
+      }
+
+      if (resp.status === 401) {
+        // Something went wrong with the refreshed access token (this should never happen)
+        throw Error('Invalid refreshed access token');
+      }
+
+      showWarningMessage(
+        opts.errorMessage
+          ? opts.errorMessage
+          : `Error while communicating with backend: ${resp.status} ${resp.statusText}`
+      );
+      return null;
+    } catch (err) {
+      // Refresh token flow or retried API call 401s again. Force logout in non-AssessmentWorkspace routes.
+      const assessmentPathRegex = /\/courses\/\d+\/[a-zA-Z]+\/\d+\/\d+/;
+      const isAssessmentUrl = !!window.location.pathname.match(assessmentPathRegex);
+
+      const userSessionExpiredNotificationKey = 'userSessionExpired';
+      if (isAssessmentUrl) {
+        showWarningMessage(
+          <div>
+            User session expired. Please copy your work and{' '}
+            <Button onClick={() => (window.location.pathname = `/login`)}>Relogin</Button>
```

Comment:
cant use react router navigate as the notification toaster is not rendered in a react router context